src/traces/sankey/render.js

```diff
@@ -0,0 +1,568 @@
+            var uniqueKeys = {};
+            persistOriginalPlace(nodes);
+            return nodes
+                .filter(function(n) {return n.visible && n.value;})
+                .map(nodeModel.bind(null, uniqueKeys, d));
+        }, keyFun);
+
+    sankeyNode.enter()
+        .append('g')
+        .classed('sankeyNode', true)
+        .call(updateNodePositions)
+        .call(attachPointerEvents, sankey, callbacks.nodeEvents);
+
+    sankeyNode
+        .call(attachDragHandler, sankeyLink, callbacks); // has to be here as it binds sankeyLink
+
+    sankeyNode.transition()
+        .ease(c.ease).duration(c.duration)
+        .call(updateNodePositions);
+
+    sankeyNode.exit().transition()
+        .ease(c.ease).duration(c.duration)
+        .style('opacity', 0)
+        .remove();
+
+    var nodeRect = sankeyNode.selectAll('.nodeRect')
+        .data(repeat);
+
+    nodeRect.enter()
+        .append('rect')
+        .classed('nodeRect', true)
+        .style('stroke-width', 0.5)
+        .call(Color.stroke, 'rgba(0, 0, 0, 1)')
+        .call(sizeNode);
+
+    nodeRect
+        .style('fill', function(d) {return d.tinyColorHue;})
+        .style('fill-opacity', function(d) {return d.tinyColorAlpha;});
+
+    nodeRect.transition()
+        .ease(c.ease).duration(c.duration)
+        .call(sizeNode);
+
+    var nodeCapture = sankeyNode.selectAll('.nodeCapture')
+        .data(repeat);
+
+    nodeCapture.enter()
+        .append('rect')
+        .classed('nodeCapture', true)
+        .style('fill-opacity', 0);
+
+    nodeCapture
+        .attr('x', function(d) {return d.zoneX;})
+        .attr('y', function(d) {return d.zoneY;})
+        .attr('width', function(d) {return d.zoneWidth;})
+        .attr('height', function(d) {return d.zoneHeight;});
+
+    var nodeLabelGuide = sankeyNode.selectAll('.nodeLabelGuide')
+        .data(repeat);
+
+    nodeLabelGuide.enter()
+        .append('path')
+        .classed('nodeLabelGuide', true)
+        .attr('id', function(d) {return d.uniqueNodeLabelPathId;});
+
+    nodeLabelGuide
+        .transition()
+        .ease(c.ease).duration(c.duration)
+        .attr('d', function(d) {
+            return d3.svg.line()([
+                [d.horizontal ? (d.left ? -d.sizeAcross : d.visibleWidth + c.nodeTextOffsetHorizontal) : c.nodeTextOffsetHorizontal, d.labelY],
+                [d.horizontal ? (d.left ? - c.nodeTextOffsetHorizontal : d.sizeAcross) : d.visibleWidth - c.nodeTextOffsetHorizontal, d.labelY]
+            ]);});
+
+    var nodeLabel = sankeyNode.selectAll('.nodeLabel')
+        .data(repeat);
+
+    nodeLabel.enter()
+        .append('text')
+        .classed('nodeLabel', true)
+        .style('user-select', 'none')
+        .style('cursor', 'default')
+        .style('fill', 'black');
+
+    nodeLabel
+        .style('text-shadow', function(d) {
+            return d.horizontal ? '-1px 1px 1px #fff, 1px 1px 1px #fff, 1px -1px 1px #fff, -1px -1px 1px #fff' : 'none';
```

Comment:
`#fff` here should depend on the font color - ie `Color.contrast(d.textFont.color)` I guess?

Incidentally, I really like this text shadow - the version we have in polar doesn't have the blur but this looks nicer IMO.